### DISCUSSION TITLE
[Proposal] Add "suppress" operator in try-catch

### DISCUSSION THREAD

**Discussion by ds1709** [dotnet/csharplang#8801] [Language Ideas]
# Summary
In try-catch code block user can suppress exception explicitly with `suppress` operator.
# Motivation
Quite often there is a need to suppress some exception in try-catch, especially `OperationCanceledException`. For this we need to create empty `catch` block. It may looks ambiguously, should it be suppressed or just somone forgot to handle it. Introducing `suppress` operator indicate explicitly exception must be suppresed.
```csharp

// Without suppress operator
try
{
    DoSomeWork();
}
catch (OperationCanceledException)
{
}
catch (SqlException sqlEx) when sqlEx.ErrorCode == 123
{
}
catch (Exception ex)
{
    ShowError(ex.Message);
}
finally
{
    Finalize();
}

// With suppress operator
try
{
    DoSomeWork();
}
suppress (OperationCanceledException)
suppress (SqlException sqlEx) when sqlEx.ErrorCode == 123
catch (Exception ex)
{
    ShowError(ex.Message);
}
finally
{
    Finalize();
}
```

**Comment by YairHalberstadt** [comment]
There's a simple solution - add a comment `// suppressed`.

Most analysers which warn on empty catch blocks, don't warn if there's a comment.

**Comment by yaakov-h** [comment]
See previous similar discussion at https://github.com/dotnet/csharplang/issues/220.

**Comment by ronnygunawan** [comment]
Might as well propose a rethrow keyword
```csharp
try
{
    DoSomeWork();
}
rethrow (OperationCanceledException)
suppress (SqlException sqlEx) when sqlEx.ErrorCode == 123
catch (Exception ex)
{
   ShowError(ex.Message);
}
finally
{
    Finalize();
}
```

**Comment by yaakov-h** [comment]
That would be a separate proposal, if you feel the need for such a thing.

**Comment by TonyValenti** [comment]
I have an ex.Ignore extension method that does nothing and is conditional debug

**Comment by Thaina** [comment]
Didn't it just

```C#
try
{
    DoSomeWork();
}
catch(SqlException sqlEx) when sqlEx.ErrorCode == 123 {}
catch (Exception ex) when !(e is OperationCanceledException)
{
   ShowError(ex.Message);
}
finally
{
    Finalize();
}
```

Or maybe

```C#
try
{
    DoSomeWork();
}
catch (Exception ex) when !(ex is OperationCanceledException) && !(ex is SqlException sqlEx && sqlEx.ErrorCode == 123)
{
   ShowError(ex.Message);
}
finally
{
    Finalize();
}

// pattern matching?

try
{
    DoSomeWork();
}
catch (Exception ex) when ex not OperationCanceledException or SqlException with { ErrorCode : 123 }
{
   ShowError(ex.Message);
}
finally
{
    Finalize();
}
```

???

I know `rethrow` and `suppress` is cleaner but I don't see it worth enough for introduce new keyword

Maybe `rethrow` could be allowed with this syntax instead

```C#
try
{
    DoSomeWork();
}
catch (OperationCanceledException) throw; // can shorthand the throw in place of when, and it will work like your rethrow, should not push new stack
catch (Exception ex) when !(ex is SqlException sqlEx && sqlEx.ErrorCode == 123)
{
   ShowError(ex.Message);
}
finally
{
    Finalize();
}
```

**Comment by theraot** [comment]
This is how I currently deal with this:

    try
    {
        DoSomeWork();
    }
    catch (OperationCanceledException exc)
    {
        _ = exc;
    }

My Intent was expressed. I did not forget to write the catch block, I clearly want to *discard* the exception. Also all the analysis tools I've been using are happy with this, and I have plenty. *Also no "do nothing" method with conditional compilation required.*
